File: tests/cross-substrate/gf-01/a2a/generate_ma_pack.py
#!/usr/bin/env python3
"""
A2A Multi-Agent Pack Generator (Official Substrate)
Scenario: gf-01-multi-agent-lifecycle

Uses pure Python for deterministic output (no external A2A calls)

Determinism Locks:
- Fixed timestamp: 2026-01-14T00:00:00Z
- Seeded UUID: uuid5(NAMESPACE, scenario_id)
- Sorted JSON output
- No absolute paths
"""

import json
import hashlib
import os
import sys
import argparse
import uuid

# Determinism constants
FIXED_TIMESTAMP = "2026-01-14T00:00:00Z"
SCENARIO_ID = "gf-01-multi-agent-lifecycle"
RUN_ID = "gf-01-ma-a2a-official-v0.7.2"
NAMESPACE = uuid.UUID("a2a00000-1234-5678-1234-567812345678")

# Agent definitions (A2A agent-to-agent pattern)
AGENTS = [
    {"agent_id": "agent_a", "role": "coordinator"},
    {"agent_id": "agent_b", "role": "executor"}
]

def seeded_uuid(name: str) -> str:
    """Generate deterministic UUID from name."""
    return str(uuid.uuid5(NAMESPACE, name))

def write_json(path: str, data: dict):
    """Write JSON with sorted keys for determinism."""
    os.makedirs(os.path.dirname(path), exist_ok=True)
    with open(path, "w") as f:
        json.dump(data, f, indent=2, sort_keys=True)
        f.write("\n")

def generate_context():
    """Generate context.json - A2A message pattern."""
    return {
        "context_id": seeded_uuid(f"{RUN_ID}-context"),
        "created_at": FIXED_TIMESTAMP,
        "created_by": "agent_a",
        "project_id": seeded_uuid(f"{RUN_ID}-project"),
        "requirements": {
            "agents_required": 2,
            "handoff_required": True,
            "summary": "Multi-agent lifecycle with A2A agent-to-agent message passing pattern"
        }
    }

def generate_plan(context_id: str):
    """Generate plan.json with A2A message steps."""
    steps = [
        {"action": "initialize_agent", "agent": "agent_a", "step_id": "1"},
        {"action": "create_task_message", "agent": "agent_a", "step_id": "2"},
        {"action": "send_to_agent_b", "agent": "agent_a", "step_id": "3"},
        {"action": "receive_message", "agent": "agent_b", "step_id": "4"},
        {"action": "process_task", "agent": "agent_b", "step_id": "5"},
        {"action": "send_result", "agent": "agent_b", "step_id": "6"}
    ]
    
    return {
        "approach": "Agent A creates task and sends to Agent B via A2A message",
        "context_ref": context_id,
        "created_at": FIXED_TIMESTAMP,
        "created_by": "agent_a",
        "plan_id": seeded_uuid(f"{RUN_ID}-plan"),
        "steps": steps
    }

def generate_trace(context_id: str, plan_id: str):
    """Generate trace.json with A2A execution summary."""
    return {
        "agent_summaries": [
            {"agent_id": "agent_a", "artifacts_created": ["context.json", "plan.json"], "events_count": 4},
            {"agent_id": "agent_b", "artifacts_created": ["trace.json"], "events_count": 4}
        ],
        "completed_by": "agent_b",
        "created_at": FIXED_TIMESTAMP,
        "execution_summary": {
            "handoffs": 1,
            "outcome": "success",
            "total_agents": 2,
            "total_events": 8
        },
        "plan_ref": plan_id,
        "trace_id": seeded_uuid(f"{RUN_ID}-trace")
    }

def generate_timeline_events(plan_id: str):
    """Generate timeline events (NDJSON format) - A2A message pattern."""
    events = [
        {"agent_id": "agent_a", "event_id": "evt-001", "role": "coordinator", "ts": FIXED_TIMESTAMP, "type": "agent.init"},
        {"agent_id": "agent_a", "artifact_ref": "artifacts/context.json", "artifact_type": "context", "event_id": "evt-002", "ts": FIXED_TIMESTAMP, "type": "artifact.create"},
        {"agent_id": "agent_a", "artifact_ref": "artifacts/plan.json", "artifact_type": "plan", "event_id": "evt-003", "ts": FIXED_TIMESTAMP, "type": "artifact.create"},
        {"agent_id": "agent_a", "context_ref": "artifacts/plan.json", "event_id": "evt-004", "from_agent": "agent_a", "payload": {"task": "process_request"}, "to_agent": "agent_b", "ts": FIXED_TIMESTAMP, "type": "handoff"},
        {"agent_id": "agent_b", "event_id": "evt-005", "received_from": "agent_a", "role": "executor", "ts": FIXED_TIMESTAMP, "type": "agent.init"},
        {"agent_id": "agent_b", "event_id": "evt-006", "task_ref": "process_request", "ts": FIXED_TIMESTAMP, "type": "task.start"},
        {"agent_id": "agent_b", "artifact_ref": "artifacts/trace.json", "artifact_type": "trace", "event_id": "evt-007", "ts": FIXED_TIMESTAMP, "type": "artifact.create"},
        {"agent_id": "agent_b", "event_id": "evt-008", "outcome": "success", "ts": FIXED_TIMESTAMP, "type": "agent.complete"}
    ]
    return events

def generate_manifest():
    """Generate pack manifest."""
    return {
        "created_at": FIXED_TIMESTAMP,
        "generator": {
            "name": "a2a-multi-agent-generator",
            "version": "0.7.2"
        },
        "multi_agent": {
            "agent_count": 2,
            "agents": AGENTS,
            "handoff_count": 1
        },
        "pack_id": RUN_ID,
        "protocol_version": "1.0.0",
        "scenario_id": SCENARIO_ID,
        "substrate": "a2a",
        "substrate_version": "0.2.1"
    }

def generate_sha256sums(pack_dir: str) -> str:
    """Generate sha256sums.txt for integrity verification."""
    files_to_hash = []
    for root, dirs, files in os.walk(pack_dir):
        for f in sorted(files):
            if f in ("sha256sums.txt", "pack_root_hash.txt"):
                continue
            filepath = os.path.join(root, f)
            relpath = os.path.relpath(filepath, pack_dir)
            files_to_hash.append((relpath, filepath))
    
    files_to_hash.sort(key=lambda x: x[0])
    
    lines = []
    for relpath, filepath in files_to_hash:
        with open(filepath, "rb") as f:
            h = hashlib.sha256(f.read()).hexdigest()
        lines.append(f"{h}  {relpath}")
    
    os.makedirs(os.path.join(pack_dir, "integrity"), exist_ok=True)
    sums_path = os.path.join(pack_dir, "integrity", "sha256sums.txt")
    with open(sums_path, "w") as f:
        f.write("\n".join(lines) + "\n")
    
    return sums_path

def main():
    parser = argparse.ArgumentParser(description="A2A Multi-Agent Pack Generator")
    parser.add_argument("--out", default="pack", help="Output directory for pack")
    args = parser.parse_args()
    
    pack_dir = args.out
    
    print(f"=== A2A Multi-Agent Pack Generator (v0.7.2) ===\n")
    print(f"Run ID: {RUN_ID}")
    print(f"Output: {pack_dir}\n")
    
    # Generate artifacts
    context = generate_context()
    write_json(f"{pack_dir}/artifacts/context.json", context)
    print(f"✓ Generated context.json (id: {context['context_id'][:8]}...)")
    
    plan = generate_plan(context["context_id"])
    write_json(f"{pack_dir}/artifacts/plan.json", plan)
    print(f"✓ Generated plan.json (id: {plan['plan_id'][:8]}..., {len(plan['steps'])} steps)")
    
    trace = generate_trace(context["context_id"], plan["plan_id"])
    write_json(f"{pack_dir}/artifacts/trace.json", trace)
    print(f"✓ Generated trace.json (id: {trace['trace_id'][:8]}...)")
    
    # Generate manifest
    manifest = generate_manifest()
    write_json(f"{pack_dir}/manifest.json", manifest)
    print(f"✓ Generated manifest.json")
    
    # Generate timeline (NDJSON)
    events = generate_timeline_events(plan["plan_id"])
    os.makedirs(f"{pack_dir}/timeline", exist_ok=True)
    with open(f"{pack_dir}/timeline/events.ndjson", "w") as f:
        for event in events:
            f.write(json.dumps(event, sort_keys=True) + "\n")
    print(f"✓ Generated timeline/events.ndjson ({len(events)} events)")
    
    # Generate integrity hashes
    sums_path = generate_sha256sums(pack_dir)
    print(f"✓ Generated integrity/sha256sums.txt")
    
    # Calculate pack_root_hash
    with open(sums_path, "rb") as f:
        pack_root_hash = hashlib.sha256(f.read()).hexdigest()
    
    # Write pack_root_hash to file
    with open(f"{pack_dir}/pack_root_hash.txt", "w") as f:
        f.write(pack_root_hash + "\n")
    
    print(f"\n✅ Pack generated successfully!")
    print(f"   pack_root_hash: {pack_root_hash}")
    
    return pack_root_hash

if __name__ == "__main__":
    main()

File: tests/cross-substrate/gf-01/mcp/generate_ma_pack.mjs
#!/usr/bin/env node
/**
 * MCP Multi-Agent Pack Generator (Official Substrate)
 * Scenario: gf-01-multi-agent-lifecycle
 * 
 * Uses pure JS for deterministic output (no external MCP SDK calls)
 * 
 * Determinism Locks:
 * - Fixed timestamp: 2026-01-14T00:00:00Z
 * - Seeded UUIDs via deterministic hash
 * - Sorted JSON output
 * - No absolute paths
 */

import fs from 'fs';
import path from 'path';
import crypto from 'crypto';

// Determinism constants
const FIXED_TIMESTAMP = "2026-01-14T00:00:00Z";
const SCENARIO_ID = "gf-01-multi-agent-lifecycle";
const RUN_ID = "gf-01-ma-mcp-official-v0.7.2";
const NAMESPACE = "mcp00000-1234-5678-1234-567812345678";

// Agent definitions (MCP server-client pattern)
const AGENTS = [
    { agent_id: "server", role: "coordinator" },
    { agent_id: "client", role: "executor" }
];

function seededUuid(name) {
    const hash = crypto.createHash('sha256').update(NAMESPACE + name).digest('hex');
    return `${hash.slice(0, 8)}-${hash.slice(8, 12)}-${hash.slice(12, 16)}-${hash.slice(16, 20)}-${hash.slice(20, 32)}`;
}

function writeJson(filePath, data) {
    fs.mkdirSync(path.dirname(filePath), { recursive: true });
    fs.writeFileSync(filePath, JSON.stringify(data, Object.keys(data).sort(), 2) + '\n');
}

function generateContext() {
    return {
        context_id: seededUuid(`${RUN_ID}-context`),
        created_at: FIXED_TIMESTAMP,
        created_by: "server",
        project_id: seededUuid(`${RUN_ID}-project`),
        requirements: {
            agents_required: 2,
            handoff_required: true,
            summary: "Multi-agent lifecycle with MCP server-client tool invocation pattern"
        }
    };
}

function generatePlan(contextId) {
    const steps = [
        { action: "initialize_server", agent: "server", step_id: "1" },
        { action: "expose_tools", agent: "server", step_id: "2" },
        { action: "handoff_to_client", agent: "server", step_id: "3" },
        { action: "receive_tool_list", agent: "client", step_id: "4" },
        { action: "invoke_tool", agent: "client", step_id: "5" },
        { action: "return_result", agent: "client", step_id: "6" }
    ];

    return {
        approach: "Server exposes tools, hands off to client for tool invocation",
        context_ref: contextId,
        created_at: FIXED_TIMESTAMP,
        created_by: "server",
        plan_id: seededUuid(`${RUN_ID}-plan`),
        steps
    };
}

function generateTrace(contextId, planId) {
    return {
        agent_summaries: [
            { agent_id: "server", artifacts_created: ["context.json", "plan.json"], events_count: 4 },
            { agent_id: "client", artifacts_created: ["trace.json"], events_count: 4 }
        ],
        completed_by: "client",
        created_at: FIXED_TIMESTAMP,
        execution_summary: {
            handoffs: 1,
            outcome: "success",
            total_agents: 2,
            total_events: 8
        },
        plan_ref: planId,
        trace_id: seededUuid(`${RUN_ID}-trace`)
    };
}

function generateTimelineEvents(planId) {
    return [
        { agent_id: "server", event_id: "evt-001", role: "coordinator", ts: FIXED_TIMESTAMP, type: "agent.init" },
        { agent_id: "server", artifact_ref: "artifacts/context.json", artifact_type: "context", event_id: "evt-002", ts: FIXED_TIMESTAMP, type: "artifact.create" },
        { agent_id: "server", artifact_ref: "artifacts/plan.json", artifact_type: "plan", event_id: "evt-003", ts: FIXED_TIMESTAMP, type: "artifact.create" },
        { agent_id: "server", context_ref: "artifacts/plan.json", event_id: "evt-004", from_agent: "server", payload: { task: "invoke_tools" }, to_agent: "client", ts: FIXED_TIMESTAMP, type: "handoff" },
        { agent_id: "client", event_id: "evt-005", received_from: "server", role: "executor", ts: FIXED_TIMESTAMP, type: "agent.init" },
        { agent_id: "client", event_id: "evt-006", task_ref: "invoke_tools", ts: FIXED_TIMESTAMP, type: "task.start" },
        { agent_id: "client", artifact_ref: "artifacts/trace.json", artifact_type: "trace", event_id: "evt-007", ts: FIXED_TIMESTAMP, type: "artifact.create" },
        { agent_id: "client", event_id: "evt-008", outcome: "success", ts: FIXED_TIMESTAMP, type: "agent.complete" }
    ];
}

function generateManifest() {
    return {
        created_at: FIXED_TIMESTAMP,
        generator: {
            name: "mcp-multi-agent-generator",
            version: "0.7.2"
        },
        multi_agent: {
            agent_count: 2,
            agents: AGENTS,
            handoff_count: 1
        },
        pack_id: RUN_ID,
        protocol_version: "1.0.0",
        scenario_id: SCENARIO_ID,
        substrate: "mcp",
        substrate_version: "2024-11-05"
    };
}

function generateSha256sums(packDir) {
    const files = [];
    function walk(dir) {
        const entries = fs.readdirSync(dir, { withFileTypes: true });
        for (const entry of entries) {
            const fullPath = path.join(dir, entry.name);
            if (entry.isDirectory()) {
                walk(fullPath);
            } else if (!['sha256sums.txt', 'pack_root_hash.txt'].includes(entry.name)) {
                files.push(fullPath);
            }
        }
    }
    walk(packDir);
    files.sort();

    const lines = [];
    for (const file of files) {
        const content = fs.readFileSync(file);
        const hash = crypto.createHash('sha256').update(content).digest('hex');
        const relPath = path.relative(packDir, file);
        lines.push(`${hash}  ${relPath}`);
    }

    const sumsPath = path.join(packDir, 'integrity', 'sha256sums.txt');
    fs.mkdirSync(path.dirname(sumsPath), { recursive: true });
    fs.writeFileSync(sumsPath, lines.join('\n') + '\n');
    return sumsPath;
}

function main() {
    const args = process.argv.slice(2);
    const outIdx = args.indexOf('--out');
    const packDir = outIdx >= 0 ? args[outIdx + 1] : 'pack';

    console.log(`=== MCP Multi-Agent Pack Generator (v0.7.2) ===\n`);
    console.log(`Run ID: ${RUN_ID}`);
    console.log(`Output: ${packDir}\n`);

    // Generate artifacts
    const context = generateContext();
    writeJson(`${packDir}/artifacts/context.json`, context);
    console.log(`✓ Generated context.json (id: ${context.context_id.slice(0, 8)}...)`);

    const plan = generatePlan(context.context_id);
    writeJson(`${packDir}/artifacts/plan.json`, plan);
    console.log(`✓ Generated plan.json (id: ${plan.plan_id.slice(0, 8)}..., ${plan.steps.length} steps)`);

    const trace = generateTrace(context.context_id, plan.plan_id);
    writeJson(`${packDir}/artifacts/trace.json`, trace);
    console.log(`✓ Generated trace.json (id: ${trace.trace_id.slice(0, 8)}...)`);

    // Generate manifest
    const manifest = generateManifest();
    writeJson(`${packDir}/manifest.json`, manifest);
    console.log(`✓ Generated manifest.json`);

    // Generate timeline (NDJSON)
    const events = generateTimelineEvents(plan.plan_id);
    fs.mkdirSync(`${packDir}/timeline`, { recursive: true });
    const ndjson = events.map(e => JSON.stringify(e, Object.keys(e).sort())).join('\n') + '\n';
    fs.writeFileSync(`${packDir}/timeline/events.ndjson`, ndjson);
    console.log(`✓ Generated timeline/events.ndjson (${events.length} events)`);

    // Generate integrity hashes
    const sumsPath = generateSha256sums(packDir);
    console.log(`✓ Generated integrity/sha256sums.txt`);

    // Calculate pack_root_hash
    const sumsContent = fs.readFileSync(sumsPath);
    const packRootHash = crypto.createHash('sha256').update(sumsContent).digest('hex');
    fs.writeFileSync(`${packDir}/pack_root_hash.txt`, packRootHash + '\n');

    console.log(`\n✅ Pack generated successfully!`);
    console.log(`   pack_root_hash: ${packRootHash}`);
}

main();
